src/app/canvas/canvas/canvas.component.ts

```diff
@@ -43,68 +46,59 @@ export class CanvasComponent {
   onMousewheel(event: WheelEvent) {
     event.preventDefault();
     const zoomScale = 1 + (event.deltaY * -0.002);
-    const zoomPnt = {
-      x: event.clientX - this.getBounds().left,
-      y: event.clientY - this.getBounds().top
-    };
-    this.canvasService.zoom(zoomPnt, zoomScale);
+    const zoomPnt = new ViewportCoord(
+      event.clientX - this.getBounds().left,
+      event.clientY - this.getBounds().top
+    );
+    this.canvasInteractionService.zoom(zoomPnt, zoomScale);
   }
 
   onMousedown(event: MouseEvent) {
     if (this.eventTargetIsCanvas(event) &&
         EventUtils.eventIsFromPrimaryButton(event)) {
-      this.panning = true;
-      this.lastPanPnt = {
-        x: event.clientX - this.getBounds().left,
-        y: event.clientY - this.getBounds().top
-      };
+      this.canvasDrag.start(new ViewportCoord(
+        event.clientX - this.getBounds().left,
+        event.clientY - this.getBounds().top
+      ));
     }
   }
 
   // Put mousemove on document to allow panning outside of canvas
   @HostListener('document:mousemove', ['$event'])
   onMousemove(event: MouseEvent) {
-    // TODO(eyuelt): We know lastPanPnt is defined if panning is true but we
-    // shouldn't have these separate fields implicitly tied together like this.
-    // What if somehow they get out of sync. Maybe create a CanvasPanner class?
-    if (this.panning && this.lastPanPnt !== null) {
+    if (this.canvasDrag.isInProgress()) {
```

Comment:
Do you have this code running in a staging environment somewhere for me to play around with it?